rn79: Restore Discover feature

jorbuedo opened the pull request:
Fix modal footers, browser context, and resolve all type/lint issues

- Fix modal system to properly render footers and titles
- Add BrowserProvider to DiscoverNavigator to resolve 'missing init' errors
- Fix all type and lint issues in Discover feature files
- Restore dApp connection and verification modals
- Fix string references and component imports
- Update navigation and tab structure
- Resolve WebView isDev property error

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - In-app browser more resilient with fallback URL and readiness signal; enables WebView debugging in development.
  - Modals support optional titles/footers and a “continue” control for guided flows.
  - Added DApps disclaimer on the Discover list.

- UI/UX
  - Refreshed styling across Discover screens and toolbars; consistent labels and icons.
  - Localization consolidated under the Discover namespace.
  - Updated warnings/confirmations (connection, HW, single-address, unverified dApp).
  - Transaction history gets a transparent, theme-aligned header.
  - Removed top border from the wallet tab bar.

- Performance/Behavior
  - More reliable welcome prompt and DApp connection handling.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jorbuedo committed:
Restore Discover feature: Fix modal footers, browser context, and resolve all type/lint issues

- Fix modal system to properly render footers and titles
- Add BrowserProvider to DiscoverNavigator to resolve 'missing init' errors
- Fix all type and lint issues in Discover feature files
- Restore dApp connection and verification modals
- Fix string references and component imports
- Update navigation and tab structure
- Resolve WebView isDev property error

jorbuedo committed:
Fix 'Invalid URL: null' error in WebView event handling

- Add fallback URL support to useConnectWalletToWebView hook
- Add WebView ready state to prevent events before WebView is loaded
- Skip events with invalid URLs and provide better error logging
- Pass tab URL as fallback to prevent null origin errors

banklesss commented on apps/mobile/src/features/Discover/common/ChainDAppsWarning.tsx:
```diff
@@ -24,24 +24,33 @@ export const ChainDAppsWarning = () => {
   )
 }
 
+const styles = StyleSheet.create({
+  container: {
+    padding: 16,
+    backgroundColor: '#fff3cd',
+    borderColor: '#ffeaa7',
+    borderWidth: 1,
+    borderRadius: 8,
+  },
+  title: {
+    fontWeight: 'bold',
+    marginBottom: 8,
+  },
+  description: {
+    marginBottom: 8,
+  },
+})
+
```
this code slipped through!

jorbuedo replied:
Yeah, looks like it's not the only one, 1 sec

jorbuedo committed:
Replace StyleSheet with atoms in ChainDAppsWarning component

- Remove StyleSheet.create usage
- Use atoms for padding and typography (a.p_md, a.font_bold, a.pb_sm)
- Keep inline styles only for colors and border properties that don't have atom equivalents
- Remove unused useTheme import

jorbuedo committed:
Replace StyleSheet.create with atoms and inline styles in Discover feature

- Remove all StyleSheet.create usage from Discover components
- Use themed atoms (ta.*) for background colors and theme-dependent properties
- Use palette (p.*) for specific colors that don't have themed atom equivalents
- Use regular atoms (a.*) for layout, spacing, and typography
- Convert complex styles to inline style objects where appropriate
- Maintain consistent styling approach across all Discover components

jorbuedo committed:
fixes

stackchain commented on apps/mobile/src/features/Discover/common/ConfirmHWConnectionModal.tsx:
```diff
@@ -32,9 +32,9 @@ export const useConfirmHWConnectionModal = () => {
   const strings = useStrings()
   const confirmHWConnection = useCallback(
-    ({onConfirm, onClose, onCancel}: Props) => {
+    ({onConfirm, onClose: _onClose, onCancel}: Props) => {
```
i think it can be deleted
```suggestion
    ({onConfirm, onCancel}: Props) => {
```

jorbuedo replied:
I'm not so sure, I don't even know why it was deleted from the modals itself. Right now the state never changes to indicate that you have already seen the modal, because that used to be on the onClose. I didn't make the decision one way or the other because I just wanted to unlock the feature, but need to decide based on other places using it whether modals should get onClose again or not.

stackchain commented on apps/mobile/src/features/Discover/common/ConfirmHWConnectionModal.tsx:
```diff
@@ -66,11 +65,10 @@ const ConfirmHWConnectionModal = ({onConfirm}: Pick<Props, 'onConfirm'>) => {
 
-  const onSelectTransport = (transportType: TransportType) => {
-    setTransportType(transportType)
+  const onSelectTransport = (selectedTransportType: TransportType) => {
```
!error useMutation new version always throws.

stackchain commented on apps/mobile/src/features/Discover/common/hooks.tsx:
```diff
@@ -32,15 +34,61 @@ export const useConnectWalletToWebView = (
+    let webViewUrl = e.nativeEvent.url
+
+    // Handle case where webViewUrl is null or invalid
+    if (!webViewUrl || webViewUrl === 'null' || webViewUrl === 'about:blank') {
```
'null' as string is that correct?

jorbuedo replied:
Could be, being a url coming from the webview, I wouldn't be surprised. Actual null is already covered by the first condition so I think it's fine

stackchain commented on apps/mobile/src/features/Discover/common/ledger.ts:
```diff
@@ -449,7 +449,7 @@ export async function toLedgerSignRequest(
 
   const txBody = csl.FixedTransaction.fromHex(cbor).body()
-  const parsedCbor = decode(txBody.toBytes())
+  const parsedCbor = decode<Map<number, unknown>>(txBody.toBytes())
```
@lisicky can confirm 
```suggestion
  const parsedCbor = decode<Map<number, UInt8Array>>(txBody.toBytes())
```

jorbuedo replied:
Not so sure, it's called `parsedCbor` so I think it's no longer binary, and it uses:
```
parsedCbor.get(1)?.constructor?.name === 'Map'
```
So I don't think it's a UInt8Array. In any case, `unknown` is enough here for ts not to complain, and it's not really propagated anywhere else.

lisicky replied:
@stackchain @jorbuedo Instead using cbor2’s decode you can just use `serialization_format` function from transaction output.

stackchain commented on apps/mobile/src/features/Discover/common/useDAppsConnected.tsx:
```diff
@@ -1,48 +1,47 @@
 
-  return useQuery({
-    suspense: true,
+  const result = usePromise({
```
@jorbuedo usePromise works like useFocusEffect, if a the dependency can change, it needs to be wrapped in a React.useCallback(, [thingsThatCanChange]) rabbitai got this one right.  :D

jorbuedo replied:
Changed to useQuery again because we need invalidations and the suggestion from the rabbit to do them made no sense

stackchain commented on apps/mobile/src/features/Discover/common/useDAppsConnected.tsx:
```diff
@@ -1,48 +1,47 @@
+  // Since we're using usePromise instead of useQuery,
+  // invalidation is handled differently - the component will re-render
+  // and the promise will be re-executed when dependencies change
   return React.useCallback(async () => {
```
this can be deleted.

stackchain commented on apps/mobile/src/features/Discover/common/useShowWelcomeDApp.tsx:
```diff
@@ -17,23 +13,32 @@ export const useShowWelcomeDApp = () => {
   })
 
-  return [query.data, mutation.mutate] as const
+  // Update local state when the promise resolves
```
this works differently now, the client needs to call promise.resolve, the value wont change by only defining it.

jorbuedo replied:
That's... confusing, I'll wait for the new usePromise api

jorbuedo replied:
added a TODO: REVISIT

stackchain commented on apps/mobile/src/features/Discover/useCases/BrowseDapp/BrowserSearchToolbar.tsx:
```diff
@@ -52,12 +52,13 @@ export const BrowserSearchToolbar = ({
           value={searchValue}
-          placeholder={strings.searchDApps}
+          placeholder={strings.discover.searchDApps}
           placeholderTextColor={p.text_gray_low}
```
```suggestion
          placeholderTextColor={ta.text_gray_low}
```

jorbuedo replied:
That's not right, the placeholderTextColor wants a color, not an object containing {color: ...}, in this case using the palette directly is correct

stackchain commented on apps/mobile/src/features/Discover/useCases/BrowseDapp/BrowserSearchToolbar.tsx:
```diff
@@ -52,12 +52,13 @@ export const BrowserSearchToolbar = ({
           autoCapitalize="none"
           style={[
-            {color: p.text_gray_max, minHeight: 36},
+            {color: p.text_gray_max},
```
```suggestion
            color: ta.text_gray_max,
```

stackchain commented on apps/mobile/src/features/Discover/useCases/BrowseDapp/BrowserTabsBar.tsx:
```diff
@@ -33,29 +33,24 @@ export const BrowserTabsBar = () => {
+        a.py_md,
+        a.px_md,
         {
           backgroundColor: p.bg_color_max,
```
```suggestion
          backgroundColor: ta.bg_color_max,
```

stackchain commented on apps/mobile/src/features/Transactions/TxHistoryNavigator.tsx:
```diff
@@ -82,6 +82,16 @@ export const TxHistoryNavigator = () => {
                 headerRight: () => <HeaderRightHistory />,
+                headerTransparent: true,
+                headerStyle: {
+                  backgroundColor: 'transparent',
```
```suggestion
                  ...a.bg_transparent,
```

jorbuedo replied:
That one's a bit on the nose, but ok

stackchain commented on apps/mobile/src/kernel/i18n/useStrings.ts:
```diff
@@ -1793,6 +1793,32 @@ export const useStrings = () => {
       },
+
+      // Mapped strings for dynamic category mapping
+      useMappedStrings: () => {
```
maybe rename to getCDapCategory strings, all React extensions think that this is hook :D 
```suggestion
      useMappedStrings: () => {
```

jorbuedo committed:
fix modals onClose

jorbuedo committed:
Fix ModalContext stale closure bug and ModalError string access issues

- Fix ModalContext: Replace useRef-based actions with useCallback to resolve stale closure issue where closeModal was capturing initial state object
- Fix ModalError: Correct string access paths from strings.global.ledgerMessages to strings.ledgerMessages
- Remove duplicate BluetoothDisabledError check in ModalError
- Ensure onClose callbacks are properly executed when modals close

jorbuedo committed:
fix

jorbuedo commented on apps/mobile/src/features/Discover/common/useDAppsConnected.tsx:
```diff
@@ -1,48 +1,47 @@
-    select: (connections) =>
-      selectWalletConnectedOrigins(
-        connections,
+  })
```
restored to useQuery
